src/main/java/foundation/privacybydesign/sms/ratelimit/RedisRateLimit.java

```diff
@@ -0,0 +1,245 @@
+                nextTry = limit.timestamp + 10 * SECOND;
+                break;
+            case 2: // try 3: allowed after 5 minutes
+                nextTry = limit.timestamp + 5 * MINUTE;
+                break;
+            case 3: // try 4: allowed after 3 hours
+                nextTry = limit.timestamp + 3 * HOUR;
+                break;
+            case 4: // try 5: allowed after 24 hours
+                nextTry = limit.timestamp + 24 * HOUR;
+                break;
+            default:
+                throw new IllegalStateException("invalid tries count");
+        }
+        return nextTry;
+    }
+
+    @Override
+    protected synchronized void countIP(String ip, long now) {
+        long nextTry = nextTryIP(ip, now);
+        if (nextTry > now) {
+            throw new IllegalStateException("counting rate limit while over the limit");
+        }
+
+        final String key = Redis.createKey(ipLimitsNamespace, ip);
+        try (var jedis = pool.getResource()) {
+            jedis.set(key, Long.toString(nextTry));
+        }
+    }
+
+    // Count the usage of this rate limit - adding to the budget for this
+    // phone number.
+    @Override
+    protected synchronized void countPhone(String phone, long now) {
+        long nextTry = nextTryPhone(phone, now);
+
+        try (var jedis = pool.getResource()) {
+            final String key = Redis.createKey(phoneLimitsNamespace, phone);
+            Limit limit = limitFromRedis(jedis, key);
+
+            if (nextTry > now) {
+                throw new IllegalStateException("counting rate limit while over the limit");
+            }
+
+            limit.tries = Math.min(limit.tries + 1, 5); // add 1, max at 5
+            // If the last usage was e.g. ≥2 days ago, we should allow them 2 tries
+            // extra tries this day.
+            long lastTryDaysAgo = (now - limit.timestamp) / DAY;
+            long bonusTries = limit.tries - lastTryDaysAgo;
+            if (bonusTries >= 1) {
+                limit.tries = (int) bonusTries;
+            }
+            limit.timestamp = now;
+
+            limitToRedis(jedis, key, limit);
+        }
+    }
+
+    @Override
+    public void periodicCleanup() {
+        cleanUpIpLimits();
+        cleanUpPhoneLimits();
+    }
+
+    /**
+     * @param jedis
+     * @param key
+     * @return null if limit was not found
+     */
+    static Limit limitFromRedis(Jedis jedis, String key) {
+        final String ts = jedis.hget(key, "timestamp");
+        final String tries = jedis.hget(key, "tries");
+
+        try {
+            return new Limit(Long.parseLong(ts), Integer.parseInt(tries));
+        } catch (NumberFormatException e) {
+            return null;
+        }
+    }
+
+    static void limitToRedis(Jedis jedis, String key, Limit limit) {
+        jedis.hset(key, "timestamp", Long.toString(limit.timestamp));
+        jedis.hset(key, "tries", Integer.toString(limit.tries));
```

Comment:
Same here as above. It is recommended to do this in a WATCH transaction.

Reply:
Good spot! I've added this.